File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def helper(tmp, t):
            if t > target:
                return
            if t == target:
                tmp = sorted(tmp)
                if tmp not in ans:
                    ans.append(tmp[:])
                return
            else:
                for i in candidates:
                    helper(tmp+[i], t+i)
        helper([], 0)
        return ans
        

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void helper(int[] candidates, int target, List<List<Integer>> answer, int index, List<Integer> tmpArray){
        if (index == -1){
            if (target == 0){
                answer.add(new ArrayList<>(tmpArray));
            }
            return;
        }
        if (target < 0)
            return;
        if (target == 0) {
            answer.add(new ArrayList<>(tmpArray));
            return;
        }
        tmpArray.add(candidates[index]);
        helper(candidates, target-candidates[index], answer, index, tmpArray);
        tmpArray.remove(tmpArray.size() - 1);
        helper(candidates, target, answer, index-1, tmpArray);
        return;
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> answer = new ArrayList<>();
        int n = candidates.length;
        helper(candidates, target, answer, n-1, new ArrayList<>());
        return answer;
    }
}